File: apps/api/app/routes/jobs.py
"""
Purpose: Expose authenticated background-job inspection, cancellation, and resume routes.
Scope: Entity-scoped list/detail reads plus operator-issued cancel and resume controls
for checkpointed worker tasks.
Dependencies: FastAPI, shared request authentication, durable job service, and
strict job API contracts.
"""

from __future__ import annotations

from typing import Annotated
from uuid import UUID

from apps.api.app.dependencies.db import DatabaseSessionDependency
from apps.api.app.dependencies.tasks import TaskDispatcherDependency
from apps.api.app.routes.request_auth import RequestAuthDependency
from fastapi import APIRouter, Depends, HTTPException, Query, status
from services.auth.service import serialize_uuid
from services.common.enums import JobStatus
from services.contracts.job_models import (
    CancelJobRequest,
    JobDetail,
    JobListResponse,
    JobSummary,
    ResumeJobRequest,
    ResumeJobResponse,
)
from services.jobs.service import JobRecord, JobService, JobServiceError

router = APIRouter(prefix="/entities/{entity_id}/jobs", tags=["jobs"])


def get_job_service(db_session: DatabaseSessionDependency) -> JobService:
    """Construct the canonical durable job service from a request-scoped DB session."""

    return JobService(db_session=db_session)


JobServiceDependency = Annotated[JobService, Depends(get_job_service)]
StatusQuery = Annotated[list[JobStatus] | None, Query(description="Optional job-status filters.")]


@router.get(
    "",
    response_model=JobListResponse,
    summary="List background jobs for one entity",
)
def list_jobs(
    entity_id: UUID,
    auth_context: RequestAuthDependency,
    job_service: JobServiceDependency,
    close_run_id: Annotated[UUID | None, Query(description="Optional close-run filter.")] = None,
    status_filter: StatusQuery = None,
) -> JobListResponse:
    """Return recent jobs visible to the current user for one entity workspace."""

    try:
        jobs = job_service.list_jobs_for_user(
            entity_id=entity_id,
            user_id=auth_context.user.id,
            close_run_id=close_run_id,
            statuses=tuple(status_filter) if status_filter else None,
        )
    except JobServiceError as error:
        raise _build_job_http_exception(error) from error

    return JobListResponse(jobs=tuple(_to_job_summary(job) for job in jobs))


@router.get(
    "/{job_id}",
    response_model=JobDetail,
    summary="Read one background job",
)
def read_job_detail(
    entity_id: UUID,
    job_id: UUID,
    auth_context: RequestAuthDependency,
    job_service: JobServiceDependency,
) -> JobDetail:
    """Return one durable background-job record with payloads and checkpoint state."""

    try:
        job = job_service.get_job_for_user(
            entity_id=entity_id,
            job_id=job_id,
            user_id=auth_context.user.id,
        )
    except JobServiceError as error:
        raise _build_job_http_exception(error) from error

    return _to_job_detail(job)


@router.post(
    "/{job_id}/cancel",
    response_model=JobDetail,
    summary="Request cancellation for one background job",
)
def cancel_job(
    entity_id: UUID,
    job_id: UUID,
    payload: CancelJobRequest,
    auth_context: RequestAuthDependency,
    job_service: JobServiceDependency,
) -> JobDetail:
    """Request cancellation and return the updated durable job state."""

    try:
        job = job_service.request_cancellation(
            entity_id=entity_id,
            job_id=job_id,
            actor_user_id=auth_context.user.id,
            reason=payload.reason,
        )
    except JobServiceError as error:
        raise _build_job_http_exception(error) from error

    return _to_job_detail(job)


@router.post(
    "/{job_id}/resume",
    response_model=ResumeJobResponse,
    status_code=status.HTTP_202_ACCEPTED,
    summary="Resume one failed, blocked, or canceled background job from its checkpoint",
)
def resume_job(
    entity_id: UUID,
    job_id: UUID,
    payload: ResumeJobRequest,
    auth_context: RequestAuthDependency,
    job_service: JobServiceDependency,
    task_dispatcher: TaskDispatcherDependency,
) -> ResumeJobResponse:
    """Create a fresh queued job from the saved checkpoint of a prior execution."""

    try:
        resumed_job = job_service.resume_job(
            dispatcher=task_dispatcher,
            entity_id=entity_id,
            job_id=job_id,
            actor_user_id=auth_context.user.id,
            reason=payload.reason,
        )
    except JobServiceError as error:
        raise _build_job_http_exception(error) from error

    return ResumeJobResponse(resumed_job=_to_job_summary(resumed_job))


def _build_job_http_exception(error: JobServiceError) -> HTTPException:
    """Convert a job-domain error into the API's structured HTTP shape."""

    return HTTPException(
        status_code=error.status_code,
        detail={
            "code": str(error.code),
            "message": error.message,
        },
    )


def _to_job_summary(job: JobRecord) -> JobSummary:
    """Translate a durable job record into the public summary contract."""

    return JobSummary(
        id=serialize_uuid(job.id),
        entity_id=serialize_uuid(job.entity_id) if job.entity_id is not None else None,
        close_run_id=serialize_uuid(job.close_run_id) if job.close_run_id is not None else None,
        document_id=serialize_uuid(job.document_id) if job.document_id is not None else None,
        task_name=job.task_name,
        queue_name=job.queue_name,
        routing_key=job.routing_key,
        status=job.status,
        retry_count=job.retry_count,
        max_retries=job.max_retries,
        attempt_count=job.attempt_count,
        failure_reason=job.failure_reason,
        blocking_reason=job.blocking_reason,
        cancellation_requested_at=job.cancellation_requested_at,
        dead_lettered_at=job.dead_lettered_at,
        resumed_from_job_id=(
            serialize_uuid(job.resumed_from_job_id)
            if job.resumed_from_job_id is not None
            else None
        ),
        started_at=job.started_at,
        completed_at=job.completed_at,
        created_at=job.created_at,
        updated_at=job.updated_at,
    )


def _to_job_detail(job: JobRecord) -> JobDetail:
    """Translate a durable job record into the public detail contract."""

    summary = _to_job_summary(job)
    return JobDetail(
        **summary.model_dump(),
        actor_user_id=serialize_uuid(job.actor_user_id) if job.actor_user_id is not None else None,
        canceled_by_user_id=(
            serialize_uuid(job.canceled_by_user_id)
            if job.canceled_by_user_id is not None
            else None
        ),
        payload=job.payload,
        checkpoint_payload=job.checkpoint_payload,
        result_payload=job.result_payload,
        failure_details=job.failure_details,
        trace_id=job.trace_id,
    )


__all__ = ["router"]

File: apps/desktop-ui/src/lib/jobs.ts
/*
Purpose: Provide same-origin background job helpers for hosted operator monitoring screens.
Scope: Job listing, detail reads, cancellation, and resume controls.
Dependencies: Browser Fetch APIs and existing `/api/entities/**` proxy routes.
*/

export type JobSummary = {
  attempt_count: number;
  blocking_reason: string | null;
  cancellation_requested_at: string | null;
  close_run_id: string | null;
  completed_at: string | null;
  created_at: string;
  dead_lettered_at: string | null;
  document_id: string | null;
  entity_id: string | null;
  failure_reason: string | null;
  id: string;
  max_retries: number;
  queue_name: string;
  resumed_from_job_id: string | null;
  retry_count: number;
  routing_key: string;
  started_at: string | null;
  status: string;
  task_name: string;
  updated_at: string;
};

export type JobDetail = JobSummary & {
  actor_user_id: string | null;
  canceled_by_user_id: string | null;
  checkpoint_payload: Record<string, unknown>;
  failure_details: Record<string, unknown> | null;
  payload: Record<string, unknown>;
  result_payload: Record<string, unknown> | null;
  trace_id: string | null;
};

export class JobApiError extends Error {
  constructor(
    message: string,
    public readonly status: number,
    public readonly code: string | null = null,
  ) {
    super(message);
    this.name = "JobApiError";
  }
}

function buildEntityProxyPath(entityId: string, pathSegments: readonly string[]): string {
  const encodedSegments = [entityId, ...pathSegments].map((segment) => encodeURIComponent(segment));
  return `/api/entities/${encodedSegments.join("/")}`;
}

async function requestJson<T>(url: string, init: RequestInit = {}): Promise<T> {
  const response = await fetch(url, {
    ...init,
    cache: "no-store",
    credentials: "include",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
      ...init.headers,
    },
  });
  const payload = await parseJsonResponse(response);
  if (!response.ok) {
    const detail = isRecord(payload) ? payload.detail : null;
    throw new JobApiError(
      isRecord(detail) && typeof detail.message === "string"
        ? detail.message
        : `Request failed with status ${response.status}.`,
      response.status,
      isRecord(detail) && typeof detail.code === "string" ? detail.code : null,
    );
  }
  return payload as T;
}

export async function listEntityJobs(
  entityId: string,
  options?: { closeRunId?: string },
): Promise<readonly JobSummary[]> {
  const searchParams = new URLSearchParams();
  if (options?.closeRunId) {
    searchParams.set("close_run_id", options.closeRunId);
  }
  const suffix = searchParams.size > 0 ? `?${searchParams.toString()}` : "";
  const payload = await requestJson<{ jobs: JobSummary[] }>(
    `${buildEntityProxyPath(entityId, ["jobs"])}${suffix}`,
    { method: "GET" },
  );
  return payload.jobs;
}

export async function readJobDetail(entityId: string, jobId: string): Promise<JobDetail> {
  return requestJson<JobDetail>(buildEntityProxyPath(entityId, ["jobs", jobId]), {
    method: "GET",
  });
}

export async function cancelJob(entityId: string, jobId: string, reason: string): Promise<void> {
  await requestJson(buildEntityProxyPath(entityId, ["jobs", jobId, "cancel"]), {
    method: "POST",
    body: JSON.stringify({ reason }),
  });
}

export async function resumeJob(entityId: string, jobId: string, reason: string): Promise<void> {
  await requestJson(buildEntityProxyPath(entityId, ["jobs", jobId, "resume"]), {
    method: "POST",
    body: JSON.stringify({ reason }),
  });
}

async function parseJsonResponse(response: Response): Promise<unknown> {
  const text = await response.text();
  if (text.length === 0) {
    return null;
  }

  try {
    return JSON.parse(text) as unknown;
  } catch {
    return null;
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}
